### PR TITLE
[GraphBolt] add abstract Dataset

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
@dgl-bot

File: python/dgl/graphbolt/dataset.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    def validation_set(self) -> ItemSet or ItemSetDict:
+        """Return the validation set."""
+        raise NotImplementedError
+
+    def test_set(self) -> ItemSet or ItemSetDict:
+        """Return the test set."""
+        raise NotImplementedError
+
+    def graph(self) -> CSCSamplingGraph or DGLGraph:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dataset.py:46]:
```suggestion
    def graph(self) -> Object:
```

File: python/dgl/graphbolt/dataset.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    networks. We calculate the embeddings based on their respective features
+    and the graph structure, and then utilize the embeddings to optimize the
+    neural network parameters.
+
+    *Feature Storage*:
+    A key-value store which stores node/edge/graph features.
+
+    *Graph Topology*:
+    Graph topology is eihter used by the subgraph sampling algorithm to
+    generate a subgraph or used by model train directly.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dataset.py:29]: The graph topology should not used by model training directly in our graphbolt scope.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/dataset.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/dataset.py:29]: what about `DGLGraph` for graph classification? used in `MinibatchSampler`?

Comment (frozenbugs) [reply python/dgl/graphbolt/dataset.py:29]: For graph classification, graph topology of each graph should be part of TVT set instead of graph topology.

**Pull request merged**